Source/WebCore/rendering/TextBoxPainter.cpp

```diff
@@ -311,7 +311,10 @@ void TextBoxPainter::paintForegroundAndDecorations()
     if (!contentMayNeedStyledMarkedText()) {
         auto markedText = MarkedText { startPosition(), endPosition(), MarkedText::Type::Unmarked };
         auto styledMarkedText = StyledMarkedText { markedText, StyledMarkedText::computeStyleForUnmarkedMarkedText(m_renderer, m_style, m_isFirstLine, m_paintInfo) };
-        paintCompositionForeground(styledMarkedText);
+        if (m_containsComposition)
+            paintCompositionForeground(styledMarkedText);
+        else
+            paintForeground(styledMarkedText);
```

Comment:
`paintCompositionForeground` already early returns calling `paintForeground` in that case.

If `paintCompositionForeground` isn't to be meant to be called with `!m_containsComposition`, then we should replace the early return with an assert.

If you're trying to avoid a call to `m_renderer.frame().editor()`, then we should just add a fast path in `paintCompositionForeground` when `m_containsComposition` is false.

There are 3 calls to `paintCompositionForeground` so it would be nice to be consistent.